### ISSUE TITLE
Count of tokens seen during training in Trainer

### ISSUE THREAD

**Issue by jpgard** [huggingface/transformers#27027]
### Feature request

The `Trainer` API should track and log the number of tokens seen during training.

While it sometimes could (maybe?) be possible to back out the number of tokens seen from the FLOS, or by iterating over the whole dataset, it would make a lot of sense for the Trainer API to track the number of tokens seen (and it shouldn't be necessary to completely iterate over a model's training loop just to compute the count of tokens, which is the only current implementation of any token-related metric in Trainer, [`Trainer.num_tokens()`](https://github.com/huggingface/transformers/blob/acc394c4f5e1283c19783581790b3dc3105a3697/src/transformers/trainer.py#L1180)).

This can't currently be implemented in a CallBack, because callbacks don't have access to the training data (only the trainer state).

### Motivation

Number of tokens seen is an essential metric tracked in nearly every LLM training run. It is widely considered one of the fundamental drivers of model quality (tokens seen during training is reported for nearly every major LLM release). It seems that any language model developer using Hugging Face would like to know this metric for their training runs -- it maybe even more important and useful than the FLOS, and perhaps as important as the number of gradient steps. 

In any case, it's an extremely useful number to have, and it must be tracked during training as the model consumes examples.

### Your contribution

I'm willing to contribute this but would like some guidance on the overall design first.

In particular, here's what I think a reasonable implementation would include:

- Add a `global_tokens_seen` or similar to the `TrainerState`. This would add only a single integer value to the `TrainerState`.
- Increment this during `Trainer._inner_training_loop()`
- Probably add this information to the logging outputs

What do the folks at HF think about that?

**Comment by geronimi73** [comment]
+1 
I think we need this feature

**Comment by ArthurZucker** [comment]
cc @muellerzr seems nice if we can make it efficient!

**Comment by muellerzr** [comment]
Is the `tokens_per_second` we already have as part of https://github.com/huggingface/transformers/pull/25858/files enough? Otherwise we can definitely add it :)

**Comment by jpgard** [comment]
Yeah, tokens/sec doesn't cut it for many use cases (although it is still very useful!!) -- similar to how tracking steps/sec doesn't obviate the need for a global step count.

If you can add it that would be amazing, I am sure this would be a useful feature to almost anyone training a language model. And I think there are some subtleties to how to make it work right in a distributed setting that you would probably be much better at handling....

**Comment by raghukiran1224** [comment]
agree tokens/sec/gpu is useful, but it fails to track `pad` tokens and if we were to do `SFTTrainer` with packing set to False, this number can be way off. So, we need a feature that tracks actual tokens seen.

**Comment by jpgard** [comment]
thanks @muellerzr !!
